presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java

```diff
@@ -0,0 +1,186 @@
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+public class TestRetryDriver
+{
+    private static final String ASYNC_FAILURE_MESSAGE = "Async failure";
+    private static final String SYNC_FAILURE_MESSAGE = "Sync failure";
+    private static final Duration minDuration = new Duration(10, MILLISECONDS);
+    private static final Duration maxDuration = new Duration(20, MILLISECONDS);
+    private static final Duration maxRetryTime = new Duration(60, SECONDS);
+    private static final double backOffScaleFactor = 1.5;
+    private static final int DUMMY_RESULT = 1;
+
+    private final ScheduledExecutorService retryExecutor;
+    private final ExecutorService taskExecutor;
+
+    public TestRetryDriver()
+    {
+        retryExecutor = newScheduledThreadPool(4, threadsNamed("test-retry-driver-%s"));
+        taskExecutor = newFixedThreadPool(4, threadsNamed("test-task-%s"));
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void close()
+    {
+        retryExecutor.shutdownNow();
+        taskExecutor.shutdownNow();
+    }
+
+    private enum AttemptStatus
+    {
+        SUCCESS,
+        ASYNC_FAILURE,
+        SYNC_FAILURE,
+    }
+
+    private static class MockAsyncTask
+            implements Callable<ListenableFuture<Integer>>
+    {
+        private static final long SLEEP_MILLISECONDS = 10L;
+
+        private final List<AttemptStatus> attempts;
+        private AtomicInteger nextAttempt;
+        private ListeningExecutorService executor;
+
+        public MockAsyncTask(List<AttemptStatus> attempts, ExecutorService executor)
+        {
+            this.attempts = attempts;
+            this.nextAttempt = new AtomicInteger(0);
+            this.executor = listeningDecorator(executor);
+        }
+
+        public void nextAttempt()
+        {
+            nextAttempt.getAndIncrement();
+        }
+
+        public int getNextAttempt()
+        {
+            return nextAttempt.get();
+        }
+
+        @Override
+        public String toString()
+        {
+            int attempt = getNextAttempt();
+            return format("next attempt: %s, last attempt status: %s", attempt, (attempt == 0) ? "NOT_EXECUTED" : attempts.get(attempt - 1));
+        }
+
+        @Override
+        public ListenableFuture<Integer> call()
+                throws Exception
+        {
+            int attempt = getNextAttempt();
+            assertTrue(attempt < attempts.size(), "current attempt is not specified");
+            AttemptStatus attemptStatus = attempts.get(attempt);
+
+            nextAttempt();
+
+            if (attemptStatus.equals(AttemptStatus.SYNC_FAILURE)) {
+                throw new RuntimeException(SYNC_FAILURE_MESSAGE);
+            }
+
+            return executor.submit(() -> {
```

Comment:
You don't need this `executor` to simulate a future. Use CompletableFuture.{complete,completeExceptionally}. Then you can use MoreFutures.toListenableFuture() to convert it.

Reply:
Refactored.